Step 78

tyand96 opened the pull request:
A Black-Scholes program.

bangerth commented:
@tyand96 is one of my students. I would appreciate if someone else could take a look at the program and provide feedback.

Tyler: Can you please add a newline character to the end of the `step-78.cc` file? One of the automatic checks currently complains about that. (Once this is fixed, it will likely complain about many more things -- go to this issue on github and see what the red crosses for all of the checks are about by looking at the "Details" link.)

peterrum commented:
@tyand96 Thanks for the contribution: I am looking forward to this one, half a year ago I was thinking to implement these type of equations for 6D ;) May I ask you post a PDF version of the tutorial. At this stage, I have problems to be build the documentation locally.

tyand96 commented:
[BS_Tutorial_Program.pdf](https://github.com/dealii/dealii/files/6290751/BS_Tutorial_Program.pdf)
I added the newline to the end of step-78.cc. Attached is the pdf of the introduction.

peterrum commented:
Thanks! I'll take a look tomorrow!

peterrum commented on examples/step-78/doc/intro.dox:
```diff
@@ -0,0 +1,213 @@
+\maketitle
+
+\section{Introduction}
+The Black-Scholes equation is a partial differential equation that falls a bit out of the ordinary scheme. It describes what the fair price of a "European call" stock option is. Without going into too much detail, a stock "option" is a contract one can buy from a bank that allows me, but not requires me, to buy a specific stock at a fixed price $K$ at a fixed future time $T$ in the future. The qu ...
```
Could you spit the lines after 80 characters?

peterrum commented on examples/step-78/doc/intro.dox:
```diff
@@ -0,0 +1,213 @@
+\maketitle
+
+\section{Introduction}
+The Black-Scholes equation is a partial differential equation that falls a bit out of the ordinary scheme. It describes what the fair price of a "European call" stock option is. Without going into too much detail, a stock "option" is a contract one can buy from a bank that allows me, but not requires me, to buy a specific stock at a fixed price $K$ at a fixed future time $T$ in the future. The qu ...
```
``"European call''

tyand96 replied:
Are you asking me to remove the quotes?

peterrum replied:
Latex normally expects different types of quotes (``) at the beginning.

tyand96 replied:
Ok, I changed the quotes back and fixed all the align environments. I also added the correct header tags.

peterrum commented on examples/step-78/doc/intro.dox:
```diff
@@ -0,0 +1,213 @@
+\end{align*}
+and we can use this as a reasonable boundary condition at our finite point $\smax$
+
+The second complication of the Block-Scholes equation is that we are given a final condition, and not an initial condition. This is because we know what the option is worth at time $t=T$: If the stock price at $T$ is $S<K$, then I have no incentive to use my option of buying a price $K$ because I can buy that stock for cheaper on the open market. So $V(S,T)=0$ for $S<K$. On the other hand, if at  ...
```
"In other words" is written twice.

peterrum commented on examples/step-78/doc/intro.dox:
```diff
@@ -0,0 +1,213 @@
+\end{align}
+
+
+As usual, (1) becomes $-\bold{M}V^n+\bold{M}V^{n-1}$ and (4) becomes $k_n\left[-r(1-\theta)\bold{M}V^{n-1} - \theta r\bold{M}V^n\right]$, where $\bold{M}_{i,j}=\left<\phi_i(S),\phi_j(S)\right>$, and where we have taken the liberty of denoting by $V$ not only the function $V(S)$ but also the vector of nodal values after discretization.
```
Why not round brackets in `$\bold{M}_{i,j}=\left<\phi_i(S),\phi_j(S)\right>$`?

tyand96 replied:
I changed this and all the other places that this happened to match the other tutorials.

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+{
+  using namespace dealii;
+
+  template <int dim>
+  double coefficient(const Point<dim> &p)
+  {
+    return p.square();
+  }
+
+  template <int dim>
+  Tensor<1, dim> SCoefficient(const Point<dim> &p)
+  {
+    return Tensor<1, dim>(p);
+  }
```
Are these functions needed?

tyand96 replied:
I think I had these in there with the intention of easily moving to higher dimensions later. I think for the 1D case at lease, we could get rid of these functions and calculate these directly inline where needed.

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+  class Solution : public Function<dim>
+  {
+  public:
+    Solution(double __maturity_time);
```
We normally do not use prefix "_" or "__".

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+                              const unsigned int component = 0) const;
+
+  private:
+    double _maturity_time;
```
Here as well.

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+  double Solution<dim>::value(const Point<dim> & p,
+                              const unsigned int component) const
+  {
+    double return_value = 0;
+    return_value +=
+      -std::pow(p(component), 2) - std::pow(this->get_time(), 2) + 6;
+    return return_value;
```
```suggestion
    return -std::pow(p(component), 2) - std::pow(this->get_time(), 2) + 6;
```

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+  Tensor<1, dim> Solution<dim>::gradient(const Point<dim> & p,
+                                         const unsigned int component) const
+  {
+    Tensor<1, dim> return_value;
+    return_value = Point<1>(-2 * p(component));
+    return return_value;
```
I guess you can do a similar simplification here.

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+                                           const unsigned int component) const
+  {
+    Tensor<1, dim> return_value;
+    for (int i = 0; dim; i++)
```
```suggestion
    for (int i = 0; dim; ++i)
```

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+
+    void run();
+
+    const double s_max;
+    const double maturity_time;
+    const double sigma;
+    const double r;
+    const double strike_price;
```
Comments are missing.

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+  public:
+    BlackScholes();
+
+    // enum RefinementMode
+    // {
+    //     global_refinement,
+    //     adaptive_refinement
+    // };
```
Are you planing to enable these?

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+    void setup_system();
+    void solve_time_step();
+    void solve_time_step_diffusion();
+    void refine_mesh(const unsigned int min_grid_level,
+                     const unsigned int max_grid_level);
+    void refine_grid();
```
Why do these two functions have different names?

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+    void refine_mesh(const unsigned int min_grid_level,
+                     const unsigned int max_grid_level);
+    void refine_grid();
+    void o_results();
```
```suggestion
    void o_results();
```

tyand96 replied:
I'm not sure what the change is here.

bangerth replied:
I don't know either, but maybe name the function `output_results()` like in all other programs.

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+  {
+    (void)component;
+    return -std::pow(p(component), 2) + 6;
+    // Below is the original initial condition
+    // return std::max(p(component) - s_price, 0.);
```
Could we introduce a parameter somewhere so that we can switch between original version and MMS?

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+    // return 0.0
+  }
+
+  /*
+  Now, we get to the implementation of the main class. This is the constructor,
+  which sets the various parameters as described in the introduction.
+  */
+  template <int dim>
+  BlackScholes<dim>::BlackScholes()
+    : fe(1)
+    , dof_handler(triangulation)
+    , time(0.0)
+    , timestep_number(0)
+    , theta(0.5)
+    , s_max(1.)
+    , sigma(.2)
+    , r(0.05)
+    , strike_price(0.5)
+    , maturity_time(1.)
+    , n_cycles(8)
+  {}
```
Let's move this down.

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+    , n_cycles(8)
+  {}
+
+  template <int dim>
+  InitialConditions<dim>::InitialConditions(double _s_price)
+  {
+    s_price = _s_price;
+  }
```
```suggestion
  template <int dim>
  InitialConditions<dim>::InitialConditions(const double _s_price) : s_price(_s_price)
  {
  }
````

same below

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+                                      QGauss<dim>(fe.degree + 1),
+                                      mass_matrix);
+
+    // Code to make Laplace Matrix with non constant coefficients
+    /*
+    Below is the code to create the Laplace matrix with non constant
+    coefficients. This corresponds to the matrix D in the introduction. This
+    non-constant coefficient is represented in the 'current_coefficient'
+    variable.
+    */
```
Merge. Same below.

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+    solution.reinit(dof_handler.n_dofs());
+    old_solution.reinit(dof_handler.n_dofs());
+    system_rhs.reinit(dof_handler.n_dofs());
+  }
+  /*
```
Missing gap.

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+  }
+
+  /*
+  This is somewhat unnecessary to have a function for the global refinement that
```
```suggestion
  It is somewhat unnecessary to have a function for the global refinement that
```

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+  This is somewhat unnecessary to have a function for the global refinement that
+  we do. The reason for the function, is to allow for the possibility of an
+  adaptive refinement later.
+  */
+
+  template <int dim>
```
```suggestion
```

tyand96 replied:
I'm not sure what you're saying with this. You don't want me to remove these lines do you?

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+            o_results();
+          }
+
+        // In the case of the line, the boundary id's are already what they
+        // should be
```
Is this statement needed?

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+  {
+    const unsigned int initial_global_refinement = 0; // 5 8
+
+    GridGenerator::hyper_cube(triangulation, 0.0, s_max);
```
Why not `GridGenerator::hyper_cube(triangulation, 0.0, s_max, true);`?

peterrum replied:
A general question. Would the code also work for 2 and 3D?

tyand96 replied:
I think I didn't use true here, because in 1D it always colorizes. I changed it just now though so that it would be this way if we elevated it. For now, it doesn't work in higher dimensions because I'm not implementing the general formula that counts for the correlation between assets. I would like to do this soon, but I don't think it'll be ready in the next few days.

peterrum commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,869 @@
+      BlackScholes<1> black_scholes_solver;
+      black_scholes_solver.run();
+    }
+  catch (...)
+    {
+      std::cout << "Error";
+    }
```
What do other tutorial do here? This does not look right^^

peterrum reviewed (commented):
I left some comments. The tutorial looks good already! Thanks!

bangerth commented on examples/step-78/doc/tooltip:
```diff
@@ -0,0 +1 @@
+Black-Scholes equation for stock options.
```
Can you add a newline to the end of the file?

marcfehling commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,835 @@
+    Tensor<1, dim> return_value;
+    for (int i = 0; dim; ++i)
+      {
+        return_value += -2 * (this->get_time()) * std::sin(p(i)) * Point<1>(1);
```
```suggestion
        return_value[0] += -2 * (this->get_time()) * std::sin(p(i));
```

marcfehling reviewed (changes requested):
Thank you for your contribution. I like the presentation of the tutorial, especially the main class. I also like that you used range-based for loops wherever possible.

I have just a few requests:
- Could you try to comply the overall layout more to the other tutorials steps, i.e., documentation syntax and grouping the implementation of classes.
- The `indent` checker complains that there is an issue with the documentation. I guess there is an issue with one of the LaTeX commands.
```
/home/runner/work/dealii/dealii/build/doc/doxygen/tutorial/step-78.h:1337: warning: Found unknown command `\f'
/home/runner/work/dealii/dealii/build/doc/doxygen/tutorial/step-78.h:1338: warning: Found unknown command `\f'
/home/runner/work/dealii/dealii/build/doc/doxygen/tutorial/step-78.h:1357: warning: Found unknown command `\f'
```

Could you build the doxygen documentation [locally on your machine](https://www.dealii.org/developer/readme.html#documentation)? The intro.dox, step-78.cc, and results.dox files will be basically glued together in the process, resulting in the file step-78.h. There, you can identify the exact line that is responsible for the error. Further, once the documentation of the tutorial is built successfully, could you post a pdf of it here? I would like to know how it looks like :-)

tyand96 commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,835 @@
+    gradient(const Point<dim> & p,
+             const unsigned int component = 0) const override;
+
+    Tensor<1, dim> time_deriv(const Point<dim> & p,
```
I removed this function. It turns out I didn't need it.

tyand96 commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,835 @@
+  {
+    (void)component;
+    Tensor<1, dim> return_value;
+    for (int i = 0; dim; ++i)
```
Wow, that's terrible... The reason this worked was because I never actually used this function. So this chunk of code was never run. I've since removed this function though since I discovered I didn't need it.

tyand96 commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,835 @@
+  Tensor<1, dim> Solution<dim>::gradient(const Point<dim> & p,
+                                         const unsigned int component) const
+  {
+    return Point<1>(-2 * p(component));
```
Yes, this program only works for dim=1 for now. I added the assertion into the constructor for the BlackScholes class so that the program will fail immediately if the user tries to run any other dimension.

marcfehling replied:
Could you add this assertion also to the constructor of the `Solution` class? This class forms a closed entity that should also contain all of the assumptions you make.

tyand96 commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,835 @@
+    // return 0.0
+  }
+
+  template <int dim>
+  InitialConditions<dim>::InitialConditions(double s_price)
+    : s_price(s_price)
+  {}
+
+  template <int dim>
+  LeftBoundaryValues<dim>::LeftBoundaryValues()
+  {}
+
+  template <int dim>
+  RightBoundaryValues<dim>::RightBoundaryValues(double s_price, double r)
+    : s_price(s_price)
+    , r(r)
+  {}
+
+  template <int dim>
+  RightHandSide<dim>::RightHandSide(double sigma, double r)
+    : sigma(sigma)
+    , r(r)
+  {}
+
+  template <int dim>
+  Solution<dim>::Solution(double maturity_time)
+    : maturity_time(maturity_time)
+  {}
```
I'm not sure what you mean here with the mixing of member functions. I am defining some member variables that have the same name such as 'r'. Is this what you are referencing to?

marcfehling replied:
These are all definition of constructors for different classes in lines 283 to 307.

For tutorials, we made sure that all definitions of class member functions immediately follow the class definition. So you basically have one block of code that deals with that particular class and no other to prevent confusion amongst readers. This way we do not mix up implementations for different classes.

Just move the definitions to the correct position. I mean something like this:
```
// Solution
template <int dim> class Solution {...};               // declaration of class
template <int dim> Solution<dim>::Solution(...){...}   // definition of constructor
template <int dim> Solution<dim>::value(...){...}      // definition of another member
...

// InitialConditions
template <int dim> class InitialConditions {...};
template <int dim> InitialConditions<dim>::InitialConditions(...){...}
...
```

tyand96 commented on examples/step-78/doc/results.dox:
```diff
@@ -0,0 +1,93 @@
+
+What is more interesting is the output of the convergence tables. They are outputted into the console, as well into a LaTex file. Below are the contents of these files.
+
+@f{table}[h]{
```
I looked at some of the other tutorial programs that would output a table during the program, and it looks like the standard is to have the console output on the tutorial webpage and then just mention that the program outputs a .tex file. This is what I changed in my results.dox.

marcfehling replied:
I'm fine with any way you want to integrate these tables, as long as [`doxygen` builds the documentation](https://www.dealii.org/developer/readme.html#documentation) successfully.

marcfehling replied:
I've started the github CI workflow. Let's see if the `indent` runner passes this time.

tyand96 commented:
@marcfehling I corrected almost all of the changes you suggested. The only one left is correcting the image in the results. I am still working on putting together a good image to put on the website, but once this is done I will correct that one too.

marcfehling commented:
Can you update the [code indentation](https://www.dealii.org/current/doxygen/deal.II/CodingConventions.html) with `make indent`? 

You may need to download `clang-format` first, which you can do conveniently with the `contrib/utilities/download_clang_format` script.

marcfehling commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,889 @@
+      VectorTools::compute_global_error(triangulation,
+                                        difference_per_cell,
+                                        VectorTools::H1_seminorm);
+    const QTrapez<1>  q_trapez;
+    const QIterated<dim> q_iterated(q_trapez, fe.degree * 2 + 1);
```
```suggestion
    const QTrapezoid<1>  q_trapezoid;
    const QIterated<dim> q_iterated(q_trapezoid, fe.degree * 2 + 1);
```
`QTrapez` has been flagged deprecated. That's the reason why the CI runners complain:
```
/home/runner/work/dealii/dealii/examples/step-78/step-78.cc: In member function ‘void BlackScholesSolver::BlackScholes<dim>::process_solution()’:
/home/runner/work/dealii/dealii/examples/step-78/step-78.cc:578:23: error: ‘using QTrapez = class dealii::QTrapezoid<1>’ is deprecated [-Werror=deprecated-declarations]
     const QTrapez<1>  q_trapez;
                       ^~~~~~~~
In file included from /home/runner/work/dealii/dealii/examples/step-78/step-78.cc:28:0:
/home/runner/work/dealii/dealii/include/deal.II/base/quadrature_lib.h:137:51: note: declared here
 using QTrapez DEAL_II_DEPRECATED = QTrapezoid<dim>;
                                                   ^
```

tyand96 replied:
Sorry, I made this change temporarily when I was testing that everything still worked on my machine and I forgot to change it back...

marcfehling reviewed (commented):
I took the liberty to flag most of the suggestions as resolved. If you change something in your pull request that solves a suggestion, you can flag it as resolved to keep track of the progress.

juleoc02 commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,889 @@
+
+  // @sect4{<code>BlackScholes::add_results_for_output</code>}
+
+  // This is simply the function to build the solution together. For this, we
```
"build the solution together" is a bit awkward

marcfehling commented:
I've built the tutorial [documentation locally](https://www.dealii.org/developer/readme.html#documentation) and noticed a few things regarding the proper display of the documentation. I've addressed some of them in https://github.com/tyand96/dealii/pull/1. There are still a few things I would like to ask you to fix:
- `doxygen` (or rather `mathjax`) doesn't know about the LaTeX command `\bold`. Would you please find a substitute?
- Does your code compile and run when you disable `MMS`?
- Do you actually need two vectors `old_solution` and `solution`. It appears to me that you can get rid of the former.

Further, there is a merge conflict. Once you rebase on the master branch, you will be asked to solve it.

tyand96 commented on examples/step-78/step-78.cc:
```diff
@@ -0,0 +1,898 @@
+            // convergence table data.
+            solve_time_step();
+
+            if (cycle == (n_cycles - 1))
+              {
+                add_results_for_output();
+              }
```
On lines 726-729, I do this because I need to save the initial conditions in the data_out_stack. Then in lines 837-840, I save each time-step. So I think I needed to do it twice. How did you think this should be done?

marcfehling replied:
Oh for some reason I thought they were both part of the loop. This makes sense now.

tyand96 commented:
@marcfehling I changed the instances of \bold to \bftext. It looks like mathjax should support this command. 
I ran the code with MMS turned off, and it still runs fine.
I had originally had the 'old_solution' variable in there because it made sense to me based on the equations I had (since I had V^{n-1}), but you are right, that I did not actually need it in there for the program to run correctly. I have changed this.

marcfehling commented on doc/doxygen/references.bib:
```diff
@@ -1125,6 +1125,30 @@ @article{eiwa96
+  pages={801--824},
+  year={1969},
+  publisher={Wiley Online Library}
+
```
```suggestion
}
```

marcfehling commented on examples/step-78/doc/intro.dox:
```diff
@@ -0,0 +1,363 @@
+of as the inflation rate that simply devalues all money over time). For their
+work, Black and Scholes received the Nobel Prize in Economic Sciences in 1997,
+making this the first tutorial program dealing with a problem for which someone
+has gotten a Nobel Prize \cite{black1973pricing}.
```
```suggestion
has gotten a Nobel Prize @cite black1973pricing.
```

marcfehling commented on examples/step-78/doc/intro.dox:
```diff
@@ -0,0 +1,363 @@
+
+Second, after truncating the domain, we need to ask what boundary values we
+should pose at this now finite boundary. To take care of this, we use "put-call"
+parity \cite{stoll1969relationship}. A "pull option" is one in which I am
```
```suggestion
parity @cite stoll1969relationship. A "pull option" is one in which I am
```

marcfehling reviewed (commented):
`doxygen` needs a different command for citing.

marcfehling commented:
There are still occurrences of bold in your documentation. I'll replace them.

marcfehling reviewed (approved):
/rebuild

marcfehling commented:
Please squash your commits into one with `git rebase`. Since you merged with the master branch in your commit history, this can get tricky. Let us know if you need help.

drwells commented:
I am pretty happy with the state of this tutorial but I did find a few things worth improving - see https://github.com/tyand96/dealii/pull/2.

bangerth commented:
In the interest of time, if you're reasonably happy with the state, you could merge and make your PR against the master branch.

drwells commented:
That's a good idea!
